pkg/chunk/composite_store.go

```diff
@@ -171,11 +177,17 @@ func (c compositeStore) Stop() {
 	}
 }
 
-func (c compositeStore) forStores(from, through model.Time, callback func(from, through model.Time, store Store) error) error {
+func (c compositeStore) forStores(ctx context.Context, userID string, from, through model.Time, callback func(ctx context.Context, from, through model.Time, store Store) error) error {
 	if len(c.stores) == 0 {
 		return nil
 	}
 
+	var err error
+	ctx, err = c.injectCacheGen(ctx, userID)
```

Comment:
Since we add the cache gen to the context in the query frontend we shouldn't have to inject it here.

Reply:
We would still need it because we add cache gen to keys during writes as well and not everyone would be using query frontend. Did I miss something?

Reply:
When I say `writes` I mean when the ingesters flush chunks to the store. Writes to the chunks-cache and index-cache only occur during query operations. The only cache that is accessed during `Put`/`PutOne`/`PutChunks`/`BatchWrite` operations is the write-dedup cache. Which really has no need for cachegen anyway. I'm proposing we add middleware to the query api for both the querier and the query-frontend that ensures a cache gen is provided and if not it adds one to the context. We could add a value for both the results cache and store here. Then we have no need to load a cache gen number downstream. The only code that needs to be aware of it's existence is the cache clients themselves.

Reply:
Actually I was incorrect, the chunkstore does write it's chunks using a writeBackCache. However, that code intentionally does not return an error. 

```
	if cacheErr := c.writeBackCache(ctx, chunks); cacheErr != nil {
		level.Warn(log).Log("msg", "could not store chunks in chunk cache", "err", cacheErr)
	}
```

The primary issue I have is the blanket error return on all calls to `forStore` if the cachegen number can't be retrieved and injected. Caching is by definition best effort. We should not be failing writes or reads because of a caching error.

Reply:
The error is returned only when it fails to load the gen number from the store the first time and once it loaded it should never throw an error because we serve them from memory next time. If we do not return any error we would never know something is wrong unless we go and check logs or have an alert for it. I understand that caching is best effort but the errors here are thrown while loading something from db not while reading or writing the cache entries.

Reply:
If that is the case, why return an error here at all. Why not have the cachegen fail on startup if it can't load from the DB? If we are returning an error here when the ingester is trying to flush chunks it's already too late and we will drop data.

Reply:
Further, even with we just return error, this is on the flush path, which would mean we need to have alerts to know about the flushes alert anyways.